lib/runner.go

```diff
@@ -194,79 +219,86 @@ func (r *Runner) runAllTests(testFiles []string, testFolder string) {
+	ctx, cancel := context.WithTimeout(context.Background(), r.timeout)
+	defer cancel()
+
+	testPath := filepath.Join(testFolder, testFile)
+
+	command := exec.CommandContext(ctx, testPath)
+
 	commandOutput := &bytes.Buffer{}
 	command.Stdout = commandOutput
 	command.Stderr = commandOutput
 
-	// Propagate timeout information from brain to script, via the
-	// environment of the script.
+	// Propagate timeout information from brain to script, via the environment of the script.
 	env := os.Environ()
-	env = append(env, fmt.Sprintf("TESTBRAIN_TIMEOUT=%d", int(r.Timeout.Seconds())))
+	env = append(env, fmt.Sprintf("TESTBRAIN_TIMEOUT=%d", int(r.timeout.Seconds())))
 	command.Env = env
 
 	err := command.Start()
 	if err != nil {
 		return ErrorTestResult(testFile, err)
 	}
-	var timeoutLock sync.Mutex
-	timeoutReached := false
-	timer := time.AfterFunc(r.Timeout, func() {
-		command.Process.Kill()
-		timeoutLock.Lock()
-		defer timeoutLock.Unlock()
-		timeoutReached = true
-	})
-	err = command.Wait()
-	timer.Stop()
-	timeoutLock.Lock()
-	if timeoutReached {
-		commandOutput.WriteString(fmt.Sprintf("Killed by testbrain: Timed out after %v", r.Timeout))
-	}
-	timeoutLock.Unlock()
 
-	exitCode, err := r.getErrorCode(err, command)
-	if err != nil {
-		return ErrorTestResult(testFile, err)
-	}
+	done := make(chan error)
+	go func() {
+		done <- command.Wait()
+	}()
+
+	timeout := time.After(r.timeout)
 
-	return TestResult{
+	testResult := TestResult{
 		TestFile: testFile,
-		Success:  command.ProcessState.Success(),
-		ExitCode: exitCode,
-		Output:   string(commandOutput.Bytes()),
 	}
+
+	select {
+	case <-timeout:
+		command.Process.Kill()
+		commandOutput.WriteString(fmt.Sprintf("Killed by testbrain: Timed out after %v", r.timeout))
+		testResult.Success = false
+		testResult.ExitCode = -1
+	case err = <-done:
+		testResult.ExitCode, err = r.getErrorCode(err, command)
+		if err != nil {
+			return ErrorTestResult(testFile, err)
+		}
+		testResult.Success = command.ProcessState.Success()
+	}
+	testResult.Output = string(commandOutput.Bytes())
+
+	return testResult
 }
 
 func (r *Runner) getErrorCode(err error, command *exec.Cmd) (int, error) {
 	if command.ProcessState.Success() {
 		// Not exactly necessary, since we can check Success(),
-		// but more correct than saying status code is unknown
+		// but more correct than saying status code is unknown.
 		return 0, nil
 	}
 	if err != nil {
 		exitErr, ok := err.(*exec.ExitError)
 		if ok {
-			// The program has exited with an ExitError, we can get the error code from WaitStatus
-			// See https://golang.org/pkg/os/#ProcessState.Sys
-			// Although the docs mention syscall.WaitStatus works on Unix, it seems to work on Windows too
+			// The program has exited with an ExitError, we can get the error code from WaitStatus.
+			// See https://golang.org/pkg/os/#ProcessState.Sys.
+			// Although the docs mention syscall.WaitStatus works on Unix, it seems to work on Windows
```

Comment:
The comment might need to be removed, both [unix](https://golang.org/src/syscall/syscall_linux.go?s=6754:6776#L270) and [windows](https://golang.org/src/syscall/syscall_windows.go?s=34202:34245#L888) have now implementation of this.